packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -502,16 +503,59 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
     return encodeAbiParameters(parseAbiParameters("bytes, address"), [moduleSignature, moduleAddressToUse]);
   }
 
+  async #getPaymasterUserOp(
+    userOp: Partial<UserOperationStruct>,
+    paymasterServiceData: PaymasterUserOperationDto,
+  ): Promise<Partial<UserOperationStruct>> {
+    if (paymasterServiceData.mode === PaymasterMode.SPONSORED) {
+      return this.#getPaymasterAndDataFromPaymaster(userOp, paymasterServiceData);
+    } else if (paymasterServiceData.mode === PaymasterMode.ERC20) {
+      if (paymasterServiceData.feeQuote) {
+        Logger.log("there is a feeQuote: ", paymasterServiceData.feeQuote);
+        if (!paymasterServiceData.spender || !paymasterServiceData.maxApproval) {
+          throw new Error("spender and maxApproval are required for ERC20 mode");
+        }
+        return this.#getPaymasterAndDataFromPaymaster(userOp, {
+          ...paymasterServiceData,
+          feeTokenAddress: paymasterServiceData.feeQuote.tokenAddress,
+        });
+      } else if (paymasterServiceData.preferredToken) {
+        Logger.log("there is a preferred token: ", paymasterServiceData.preferredToken);
+        const preferredToken = paymasterServiceData.preferredToken;
+        const feeQuotesResponse = await this.#getTokenFeesFromUserOp(userOp, { ...paymasterServiceData, tokenList: [preferredToken] });
+        const feeQuote = feeQuotesResponse.feeQuotes?.[0];
+        if (!feeQuote) {
+          throw new Error("Error while getting feeQuote");
+        }
+        return this.#getPaymasterAndDataFromPaymaster(userOp, { ...paymasterServiceData, feeQuote, feeTokenAddress: preferredToken });
+      } else {
+        throw new Error("FeeQuote was not provided, please call smartAccount.getTokenFees() to get feeQuote");
+      }
+    }
+    throw new Error("Invalid paymaster mode");
+  }
+
+  async #getPaymasterAndDataFromPaymaster(
```

Comment:
Why "getPaymasterAndDataFromPaymaster" and not "getPaymasterAndData" ? Is this because you want to mention that the gas estimations are coming from paymaster and not bundler ?

Reply:
Yes that's right. I thought getPaymasterAndData might be confusing on the SmartWallet class considering that's the name of the function on the paymaster, but I guess it's not that confusing. Will happily change it back

Reply:
I think "getPaymasterAndData" already speaks that is calling paymaster